protocol/consistencychecks.go

```diff
@@ -0,0 +1,328 @@
+	}
+	return str, PassedWithAProofOfInclusion
+}
+
+func (cc *ConsistencyChecks) verifyKeyLookup(requestType int,
+	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
+
+	df, ok := msg.DirectoryResponse.(*DirectoryProof)
+	if !ok {
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	ap := df.AP
+	str := df.STR
+
+	// 1. verify STR
+	if err := cc.verifySTR(cc.SavedSTR, str); err != nil {
+		return nil, err
+	}
+
+	// 2. verify Auth path
+	proofType, err := verifyAuthPath(uname, key, ap, str)
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. verify (response code, proof type) pair
+	switch {
+	case msg.Error == ErrorNameNotFound && proofType == proofOfAbsence:
+	case msg.Error == Success:
+	default:
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	// determine which kind of TB's verification we should do
+	// based on the response code and proof type
+	if msg.Error == Success && proofType == proofOfAbsence {
+		// 4.1 verify returned promise
+		if err := cc.verifyReturnedPromise(requestType,
+			msg.Error, df, uname, key); err != nil {
+			return nil, err
+		}
+	} else {
+		// 4.2 verify fulfilled promise
+		if err := cc.verifyFulfilledPromise(uname, str, ap); err != nil {
+			return nil, err
+		}
+	}
+
+	if proofType == proofOfAbsence {
+		return str, PassedWithAProofOfAbsence
+	}
+	return str, PassedWithAProofOfInclusion
+}
+
+func verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (int, error) {
+	proofType := proofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return invalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { // proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return invalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return invalidProof, ErrorBadCommitment
+		}
+		proofType = proofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return invalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, nil
+}
+
+// verifySTR checks the consistency between 2 snapshots.
+// This uses the pinning signing key in cc
+// to verify the STR's signature and should not verify
+// the hash chain using the current STR stored in cc.
+func (cc *ConsistencyChecks) verifySTR(savedSTR, str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	// TODO: clearly verify which is the client's actual expectation.
+	// See #81
+	if (str.Epoch == savedSTR.Epoch && bytes.Equal(savedSTR.Signature, str.Signature)) ||
+		(str.Epoch == savedSTR.Epoch+1 && str.VerifyHashChain(savedSTR.Signature)) {
```

Comment:
`str.Epoch == savedSTR.Epoch+1` -> does this work if the client missed some epochs?